lib/portlayer/attach/attach.go

```diff
@@ -0,0 +1,169 @@
+
+package attach
+
+import (
+	"fmt"
+	"net"
+
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/vim25/types"
+	"github.com/vmware/vic/lib/portlayer/constants"
+	"github.com/vmware/vic/lib/portlayer/exec"
+	"github.com/vmware/vic/pkg/trace"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+// Handler defines the interface
+type Handler interface {
+	Join(handle interface{}) (interface{}, error)
+	Bind(handle interface{}) (interface{}, error)
+	Unbind(handle interface{}) (interface{}, error)
+}
+
+func lookupVCHIP() (net.IP, error) {
+	// FIXME: THERE MUST BE ANOTHER WAY
+	// following is from Create@exec.go
+	ips, err := net.LookupIP(constants.ManagementHostName)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(ips) == 0 {
+		return nil, fmt.Errorf("No IP found on %s", constants.ManagementHostName)
+	}
+
+	if len(ips) > 1 {
+		return nil, fmt.Errorf("Multiple IPs found on %s: %#v", constants.ManagementHostName, ips)
+	}
+	return ips[0], nil
+}
+
+func toggle(handle *exec.Handle, connected bool) (*exec.Handle, error) {
+	// make sure a spec exists
+	handle.SetSpec(nil)
+	// get the virtual device list
+	devices := object.VirtualDeviceList(handle.Container.Config.Hardware.Device)
+
+	// select the virtual serial ports
+	serials := devices.SelectByBackingInfo((*types.VirtualSerialPortURIBackingInfo)(nil))
+	if len(serials) == 0 {
+		return nil, fmt.Errorf("Unable to bind a device with desired backing")
+	}
+	if len(serials) > 1 {
+		return nil, fmt.Errorf("Multiple matches found with desired backing")
+	}
+
+	ip, err := lookupVCHIP()
+	if err != nil {
+		return nil, err
+	}
+
+	for i := range serials {
+		serial := serials[i]
+		log.Debugf("Found a device with desired backing: %#v", serial)
+
+		c := serial.GetVirtualDevice().Connectable
+		b := serial.GetVirtualDevice().Backing.(*types.VirtualSerialPortURIBackingInfo)
+
+		serviceURI := fmt.Sprintf("tcp://%s:%d", ip, constants.SerialOverLANPort)
+
+		if b.ServiceURI == serviceURI && c.Connected == connected {
+			log.Debugf("Already in the desired state (connected: %t, serviceURI: %s)", connected, serviceURI)
+			continue
+		}
+
+		// set the values
+		log.Debugf("Setting Connected to %t", connected)
+		c.Connected = connected
+		if connected && handle.ExecConfig.Sessions[handle.ExecConfig.ID].Attach {
+			log.Debugf("Setting the start connected state to %t", connected)
+			c.StartConnected = handle.ExecConfig.Sessions[handle.ExecConfig.ID].Attach
+		}
+
+		log.Debugf("Setting ServiceURI to %s", serviceURI)
+		b.ServiceURI = serviceURI
+
+		config := &types.VirtualDeviceConfigSpec{
+			Device:    serial,
+			Operation: types.VirtualDeviceConfigSpecOperationEdit,
+		}
+		handle.Spec.DeviceChange = append(handle.Spec.DeviceChange, config)
+	}
+
+	return handle, nil
+}
+
+// Join adds network backed serial port to the caller and configures them
+func Join(h interface{}) (interface{}, error) {
+	handle, ok := h.(*exec.Handle)
+	if !ok {
+		return nil, fmt.Errorf("type assertion failed")
```

Comment:
Maybe in these cases we could include `%T` in the message, to see the actual type of the handle.